storage/aptosdb/src/pruner/state_store/mod.rs

```diff
@@ -121,96 +125,176 @@ where
-    pub fn prune_state_merkle(
+    fn get_progress(db: &DB, shard_id: Option<u8>) -> Result<Version> {
+        Ok(db
+            .get::<DbMetadataSchema>(&S::tag(shard_id))?
+            .map_or(0, |v| v.expect_version()))
+    }
+
+    fn prune_state_merkle(
         &self,
         min_readable_version: Version,
         target_version: Version,
         batch_size: usize,
-        existing_schema_batch: Option<&mut SchemaBatch>,
-    ) -> anyhow::Result<Version> {
-        assert_ne!(batch_size, 0);
-        if target_version < min_readable_version {
-            return Ok(min_readable_version);
-        }
-        let (indices, is_end_of_target_version) =
-            self.get_stale_node_indices(min_readable_version, target_version, batch_size)?;
-        if indices.is_empty() {
-            self.record_progress_impl(target_version, is_end_of_target_version);
-            Ok(target_version)
-        } else {
-            let _timer = OTHER_TIMERS_SECONDS
-                .with_label_values(&["state_merkle_pruner_commit"])
-                .start_timer();
-            let new_min_readable_version =
-                indices.last().expect("Should exist.").stale_since_version;
-
-            // Delete stale nodes.
-            if let Some(existing_schema_batch) = existing_schema_batch {
-                indices.into_iter().try_for_each(|index| {
-                    existing_schema_batch.delete::<JellyfishMerkleNodeSchema>(&index.node_key)?;
-                    existing_schema_batch.delete::<S>(&index)
-                })?;
+    ) -> Result<Version> {
+        let mut min_readable_version = min_readable_version;
+        let mut target_version_for_this_batch = min_readable_version;
+        while target_version_for_this_batch <= target_version {
+            self.record_progress(target_version_for_this_batch);
+            let next_version =
+                self.prune_top_levels(min_readable_version, target_version_for_this_batch)?;
+            self.prune_shards(target_version_for_this_batch, batch_size)?;
+            min_readable_version = target_version_for_this_batch;
+            if let Some(next_version) = next_version {
+                target_version_for_this_batch = next_version;
             } else {
-                let batch = SchemaBatch::new();
-                indices.into_iter().try_for_each(|index| {
-                    batch.delete::<JellyfishMerkleNodeSchema>(&index.node_key)?;
-                    batch.delete::<S>(&index)
-                })?;
+                break;
+            }
+        }
+        Ok(min_readable_version)
+    }
 
-                self.save_min_readable_version(new_min_readable_version, &batch)?;
+    fn prune_top_levels(
+        &self,
+        min_readable_version: Version,
+        target_version: Version,
+    ) -> Result<Option<Version>> {
+        let batch = SchemaBatch::new();
+        let next_version = self.prune_state_merkle_shard(
+            self.state_merkle_db.metadata_db(),
```

Comment:
is the top level SMT stored in metadata_db?

Reply:
Yes